elastic/kibana
# [Content management] BWC - versioned content

This PR adds BWC support in Content Management.

## Notes on implementation

* every method of the storage instance receives the request version + the latest version in the `context` object

<img width="599" alt="Screenshot 2023-03-09 at 14 24 19" src="https://user-images.githubusercontent.com/2854616/224054262-a0308e00-94fa-47f6-ad4e-942a09c175ee.png">


* It is now required to **both** register the content in the browser and on the server. This ensure that all content is versioned and no "guess" logic is in place for non versioned RPC requests
* The CM public client automatically attach the `latestVersion` information in each RPC Request `payload`. Unless a specific `version` is asked by the consumer (e.g. if for some reason the client needs a `v1` while the server is on `v2`)
* Ideally every content will use the same constant (exported from the `common` folder) to register their latest version. If they don't and by mistake they register a higher version on the client than on the browser the request will throw an error before hitting the storage layer. 
* The schema validation for the different objects that can be passed to the storage methods (`data`, `query` and `options`) has been removed. Same for the `result` schema which is not validated before sending it back to the client. It is now the responsibility of each content provider to validate the incoming data (objects) and response result. In a future PR we will provide tooling to make this validation easier.

## Test in Kibana

* start kibana with the examples `yarn start --run-examples`
* go to the content management examples and start using yhe "Todos" app
* Inspect the request: every call to the rpc crud + searc should include the latest content version

<img width="785" alt="Screenshot 2023-03-09 at 14 11 55" src="https://user-images.githubusercontent.com/2854616/224054557-164be495-c27f-4d05-b206-b53d0c07474c.png">
<img width="812" alt="Screenshot 2023-03-09 at 14 12 18" src="https://user-images.githubusercontent.com/2854616/224054559-039e3d66-d86b-416c-b816-7835d9feb50a.png">
